doc/release/upcoming_changes/15900.deprecation.rst

```diff
@@ -0,0 +1,14 @@
+Indexing errors will be reported even when selecting no elements
+----------------------------------------------------------------
+In the future, NumPy will always raise an index error when an
+integer array index contains out of bound values.
+In complex indexing operations such as::
+
+    arr = np.zeros((5, 5))
+    arr[[], [20]]
```

Comment:
Are you sure this example does what you say it does?

This case looks like it should succeed to me, since the semantics are roughly:
```python
index = [], [20]
index = tuple(np.asarray(i, np.intp) for i in index)
index = np.broadcast_arrays(*index)  # 20 is removed here.
arr[index]
```

Reply:
Yes, which is what we were doing before. The question is whether this is actually nice to ignore the index for checking?

There are two things here pandas had a test for one of them. One is this case, the other is the case below where `arr = np.zeros((5, 5, 0))` but you index only on the first two axis (thus the result also being empty). I could limit on that case I guess...

Maybe this is trickier then I thought.

Reply:
I guess pandas only ran into the second case, which is easier to reason about... I should probably try to limit this to that one as well...

Reply:
I think the only way to hit this code path is when the an axis _not_ indexed by advanced indexing is length zero, such as  `np.zeros((3, 0))[[200]]` or `np.zeros((3, 4))[[200], :0]`

Reply:
Just to be clear - if you run the code sample from this release snippet on this branch, it really does give the warning you claim?

Reply:
Yeah, its a bit tricky that I check the index before broadcasting.